Add ocean and invalid reflectance masking

gshiroma opened the pull request:
This PR adds:
- Ocean masking using a shoreline shape file (GSHHS) - new fields were added to the runconfig: `shoreline_shape_file` and `shoreline_shape_file_description`;
- Cumulative reflectance no-data masking;
- Refactor code: rename some variables to match DSWx-HLS product specs, create a function to check ancillary inputs `_check_ancillary_inputs()`;
- Update messages logged/printed to the user: mostly indentation changes.

nemo794 commented on src/proteus/dswx_hls.py:
```diff
@@ -1772,14 +1810,46 @@ def _load_hls_from_file(filename, image_dict, offset_dict, scale_dict,
+    elif fill_value is None:
+        fill_value = -9999
+
+    # create/update `invalid_ind_array`
+    # (invalid pixels are cumulative)
```
Coming in with fresh eyes, it was confusing to me that this function's primary purpose seems to be to read in a single band and update the `image_dict` with the information for just that one HLS band. (tbh, I thought this function's purpose was to read in all bands from the product!)

If you have a chance, it could be helpful to make these updates:

1) update the name of the function to something like, `_load_hls_band_from_file(..)`

2) update the input paramter `key` to being something like `band` (this is what the docstring says)

3) update the docstring's description for `image_dict` to be more descriptive, something like:
```
`Image dictionary that will store HLS band array and information about the bands.
The following keys in `image_dict` will be created/updated based on the
information found in the HLS file for the given band: `invalid_ind_array`, 
`geotransform`, `projection`, `length`, `width`.
```
(I might be missing some keys)

4) update this comment to be something like:

```suggestion
    # create/update `invalid_ind_array` with invalid pixels in this band.
    # (invalid pixels are cumulative for all bands)
```

nemo794 commented on src/proteus/dswx_hls.py:
```diff
@@ -1772,14 +1810,46 @@ def _load_hls_from_file(filename, image_dict, offset_dict, scale_dict,
+        invalid_ind_array = np.logical_or(image_dict['invalid_ind_array'],
+                                          image == fill_value)
+
+    image_dict['invalid_ind_array'] = invalid_ind_array
```
Sanity Check: Have you tested one of the problem granules that was exhibiting the "railroad tracks" with this code? Meaning, have you visually inspected outputs to ensure that the new process for generating the invalid indices is working correctly and is correctly being applied to all layers? Thanks!

gshiroma replied:
Not yet, if you could help me with that that would be great.

nemo794 commented on src/proteus/dswx_hls.py:
```diff
@@ -2734,6 +2785,175 @@ def _warp(input_file, geotransform, projection,
+            # Set up the shapefile driver 
+            shapefile_driver = ogr.GetDriverByName("ESRI Shapefile")
+
+            # TODO fix filename:
+            temp_shapefile_filename = tempfile.NamedTemporaryFile(
```
On a test run of the code, it looks like the following files are not being removed from the `scratch_dir`:

```
test_data_set/scratch_dir/tmpbhirennk.dbf
test_data_set/scratch_dir/tmpbhirennk.prj
test_data_set/scratch_dir/tmpbhirennk.shp
test_data_set/scratch_dir/tmpbhirennk.shx
```

I'm guessing this line is where the `.shp` file is being generated. Could you please update the code so that these four temp files are properly handled and cleaned up, same as the other temp files?

Thank you!

gshiroma replied:
Good catch. Thanks!

nemo794 reviewed (changes requested):
As usual, several great new features and improvements! Most of the comments are fairly minor, or sanity checks.

The only major issue I'm seeing is that the ocean mask is displaying as dark blue instead of transparent. (Project had requested that the ocean appear transparent in the geotiffs.) As you discovered and mentioned offline, this is due to a bug in gdal and is not something we can fix in the PROTEUS code. But the Project should be aware of the bug, and (hopefully) come to an agreement about how to handle ocean pixels. Preferably, this agreement would happen before merging this PR.

Thanks!

gshiroma commented on src/proteus/defaults/dswx_hls.yaml:
```diff
@@ -14,19 +14,27 @@ runconfig:
-            worldcover_description:
+            worldcover_file_description:
+
+            # Global Self-consistent, Hierarchical, High-resolution Shoreline
+            # (GSHHS) shape file
+            shoreline_shape_file:
```
@nemo794 , the file is really not required. However, we added a flag `check_ancillary_inputs_coverage` to check if all the ancillary inputs cover the tile. This flag will be on for production. If an ancillary input is not provided and `check_ancillary_inputs_coverage` is enabled, the SAS will raise an error.

gshiroma commented on src/proteus/defaults/dswx_hls.yaml:
```diff
@@ -71,6 +79,9 @@ runconfig:
             copernicus_forest_classes: [20, 111, 113, 115, 116, 121, 123, 125, 126]
 
+            # Ocean masking distance from shoreline in km
+            ocean_masking_shoreline_distance_km: 10
```
I reduced it to 5 km. When you have a chance, could you please run that over San Francisco? That would be very helpful. Thank you in advance.

gshiroma commented on src/proteus/dswx_hls.py:
```diff
@@ -3311,11 +3556,114 @@ def _crop_2d_array_all_sides(input_2d_array, margin):
+        if file_description == 'Shoreline shape file':
+            continue
+            
+        # test if tile is fully covered by the ancillary input
```
It's supposed to be the other way around. I'm updating the comments accordingly.

gshiroma commented on src/proteus/dswx_hls.py:
```diff
@@ -3311,11 +3556,114 @@ def _crop_2d_array_all_sides(input_2d_array, margin):
+
+                # test if any of the ancillary input vertices is within
+                # the tile. If so, raise an error 
+                flag_error = file_vertex.Within(tile_polygon)
+                if not flag_error:
+                    continue
```
Yes, this was intentional to make sure someone reading the code would understand that the vertex being within the tile polygon is an error. But the line has a comment above, so I'll accept your suggestion and simplify the code

gshiroma commented on src/proteus/dswx_hls.py:
```diff
@@ -3510,9 +3874,16 @@ class in the browse image. False to display PSW-Agg.
     logger.info(f'    Copernicus CGLS land cover 100m file: {landcover_file}')
+    logger.info(f'        description:'
+                f' {landcover_file_description}')
     logger.info(f'    WorldCover 10m file: {worldcover_file}')
```
These files are only required if `check_ancillary_inputs_coverage` is True. I'm adding some comments to the runconfig to make it clear.

nemo794 replied:
Thank you for the updates to the comments! Those are helpful for one of my other notes.

Here, I was asking something subtly different, perhaps I can try to clarify:

The runconfig's comments simply ask for, "ESA WorldCover map file".
In the code itself (at this line and in other places), the code gets more-specific and says, "WorldCover 10m file".

Would it be possible to harmonize all of these comments, and keep them a consistent e.g. "ESA WorldCover 10m file"? (same for the landcover file)

The resolutions (10m and 100m) are critical for the algorithm to work, so it could be good to note them in the runconfig for the users.

gshiroma commented on src/proteus/dswx_hls.py:
```diff
@@ -35,7 +35,8 @@
 
 UINT8_FILL_VALUE = 255
-FILL_VALUE_RGBA = (0, 0, 0, 255)
+OCEAN_MASKED_RGBA = (0, 0, 127, 0)
```
I talked to S. Chan, and he said he prefers to keep it as a separate class and "opaque" dark blue. If we have time, maybe I'll make it configurable through the runconfig

gshiroma commented on src/proteus/dswx_hls.py:
```diff
@@ -1772,14 +1810,46 @@ def _load_hls_from_file(filename, image_dict, offset_dict, scale_dict,
+
+
+    # read `fill_value`
+    if fill_value is None and '_FillValue' in metadata.keys():
+        fill_value = float(metadata['_FillValue'])
+    elif fill_value is None:
+        fill_value = -9999
```
No problem, thanks for checking

gshiroma commented on src/proteus/dswx_hls.py:
```diff
@@ -3311,11 +3556,114 @@ def _crop_2d_array_all_sides(input_2d_array, margin):
+    tile_min_x_utm, tile_dx_utm, _, tile_max_y_utm, _, tile_dy_utm = \
+        geotransform
+    tile_max_x_utm = tile_min_x_utm + width * tile_dx_utm
+    tile_min_y_utm = tile_max_y_utm + length * tile_dy_utm
```
I revised the code for handling the dateline case and pushed commit [3cd8ae9](https://github.com/nasa/PROTEUS/pull/36/commits/3cd8ae917e784d7877631e5148cb4447ba7929c4) with a fix and a few other improvements. Thank you for this comment, @nemo794 .

nemo794 commented on src/proteus/dswx_hls.py:
```diff
@@ -2933,9 +3057,10 @@ def _create_ocean_mask(shape_file, margin_km, scratch_dir,
             temp_shapefile_filename = tempfile.NamedTemporaryFile(
                 dir=scratch_dir, suffix='.shp').name
+            if temp_files_list is not None:
+                temp_files_list.append(temp_shapefile_filename)
```
For some reason, there are still three additional temp files created in addition to the .shp temp file. I don't know where/why/how these additional temp files are created, but their names appear to be consistent with the .shp file's name. This is a hack that seems to fix the problem:

```suggestion
                temp_files_list.append(temp_shapefile_filename)
                temp_files_list.append(temp_shapefile_filename.replace('.shp','.prj'))
                temp_files_list.append(temp_shapefile_filename.replace('.shp','.dbf'))
                temp_files_list.append(temp_shapefile_filename.replace('.shp','.shx'))
```

gshiroma replied:
Thank you for catching that, @nemo794

nemo794 reviewed (commented):
Nice work adding in those last-minute changes from Project and addressing the notes! Just a couple little more things, very minor.

gshiroma commented:
> Nice work adding in those last-minute changes from Project and addressing the notes! Just a couple little more things, very minor.

Thank you, @nemo794 , for the very helpful review. I addressed your last comments. Please, let me know if yo have any other comments or concerns.

nemo794 commented on src/proteus/dswx_hls.py:
```diff
@@ -3060,7 +3065,12 @@ def _create_ocean_mask(shapefile, margin_km, scratch_dir,
+                for extension in ['.shp', '.prj', '.dbf', '.shx']:
+                    temp_file_test = temp_shapefile_filename.replace(
+                        '.shp', extension)
+                    if not os.path.isfile(temp_file_test):
+                        continue
```
Unfortunately, this line is not going to work; the .prj, .dbf, and .shx files have not yet been created at this point in the code, and so they will not be added to the `temp_files_list`, and they are still leftover in the `scratch_dir`.

Instead, could this entire 7-line code block be moved to the end of the function (right before `return`)? I just tested this, and all temp files get removed this way.

gshiroma replied:
Good catch! Thank you, @nemo794 ! I remove that check.

nemo794 replied:
Great! Just tested again, the scratch_dir is empty now. Great!

Except now I'm thinking about this more, and I do really like the idea of doing an existence check before adding these "surprise" files to the list. Example scenario: if the caller of the function got this list back and tried to delete everything without first doing an existence check themself, they would get an error because the file does not exist.

In the current DSWx-HLS code, there is an existence check at the end of the code before deleting these files, so we will not get this error. But it's probably better practice to keep the existence check within this function, but perhaps add the files to the list at the end of this function?

gshiroma replied:
Just updated the code, @nemo794 . Please, let me know if you have any other comments or concerns. Thanks.

hfattahi reviewed (approved):
LGTM. Thanks @gshiroma

nemo794 commented:
Echo'ing @hfattahi: LGTM.

Thank you for addressing all of the comments. And really great work on the entire PR -- awesome features and improvements!

nemo794 reviewed (approved):
LGTM. Thanks @gshiroma !

gshiroma commented:
> Echo'ing @hfattahi: LGTM.
> 
> Thank you for addressing all of the comments. And really great work on the entire PR -- awesome features and improvements!

Thank you so much, @nemo794 , for your comments and for catching many issues before this PR was merged.

@hfattahi and @nemo794 , I just updated the workflow test dataset URL (zenodo reference) and merged the PR.
